app/lib/actions/candlepin/environment/set_content.rb

```diff
@@ -16,29 +16,33 @@ def existing_ids
 
         def run
           saved_cp_ids = existing_ids
-          add_ids    = input[:content_ids] - saved_cp_ids
-          delete_ids = saved_cp_ids - input[:content_ids]
-          retries = 2
-          (retries + 1).times do
-            output[:add_ids] = add_ids
-            break if add_ids.empty?
+          output[:add_ids] = input[:content_ids] - saved_cp_ids
+          output[:delete_ids] = saved_cp_ids - input[:content_ids]
+
+          until output[:add_ids].empty?
             begin
               output[:add_response] = ::Katello::Resources::Candlepin::Environment.
-                add_content(input[:cp_environment_id], add_ids)
+                add_content(input[:cp_environment_id], output[:add_ids])
               break
-            rescue RestClient::InternalServerError
-              # HACK
-              # Candlepin raises a 500 in case it gets a duplicate content id add to a environment
-              # so as pure conjecture we are using this and just guessing if its a dup id.
-              # so trying again
-              add_ids = input[:content_ids] - existing_ids
+            rescue RestClient::Conflict
+              # Candlepin raises a 409 in case it gets a duplicate content id add to an environment
+              # Since its a dup id refresh the existing ids list (which hopefully will not have the duplicate content)
+              # and try again.
+              output[:add_ids] = input[:content_ids] - existing_ids
             end
           end
 
-          output[:delete_ids]      = delete_ids
-          unless delete_ids.empty?
-            output[:delete_response] = ::Katello::Resources::Candlepin::Environment.
-                delete_content(input[:cp_environment_id], delete_ids)
+          until output[:delete_ids].empty?
```

Comment:
I would still recommend putting some counter there (let's say 5 or 10), just to avoid getting into inifinite loops: one never known, after the counter overflows, an exception should be raised